File: app/controllers/job_listings_controller.rb
class JobListingsController < ApplicationController
  before_filter :find_company
  before_filter :check_invalid_permissions_company, except: [:index, :show]
  before_filter :check_for_subscription, only: [:new, :create]
  before_filter :get_job_listing, except: [:index, :new, :create, :guide]
  before_filter :cleanup_invalid_data, only: [:create, :update_listing]
  respond_to :json, :html

  def index
    @job_listings = @company.job_listings.order('active DESC, updated_at DESC').paginate(page: params[:page], per_page: 4)
  end

  def show
  end

  def guide
    @subscription = current_company.subscription
    @has_tech_stacks = current_company.tech_stacks.exists?
  end

  def new
    unless @company.subscription.maxed_out?
      @job_listing = @company.job_listings.build
      listing_response
    else
      flash[:error] = "Can't create a new job listing.  You've reached your subscription limit."
      redirect_to company_job_listings_path(company_id: @company.id)
    end
  end

  def create
    @job_listing = @company.job_listings.build
    if @job_listing.update(@bathed_params)
      @job_listing.api_create
      respond_with(@job_listing, location: nil, status: :created)
    else
      render json: { errors: @job_listing.errors }, status: :bad_request
    end
  end

  def update_listing
    if @job_listing.update(@bathed_params)
      @job_listing.api_update
      respond_with @job_listing
    else
      render json: { errors: @job_listing.errors }, status: :bad_request
    end
  end

  def retrieve_listing
    listing_response
  end

  def destroy
    @job_listing.destroy
    @job_listing.api_destroy
    redirect_to company_job_listings_path(company_id: @company.id), notice: 'Listing has been successfully removed.'
  end

  def toggle_active
    if !@company.subscription.maxed_out? || @job_listing.active
      @job_listing.toggle_active
      redirect_to company_job_listings_path(company_id: @company.id), notice: "Successfully #{@job_listing.active? ? 'activated' : 'deactivated'} your job listing."
    else
      flash[:error] = "Can't activate another job listing.  You've reached your subscription limit"
      redirect_to company_job_listings_path(company_id: @company.id)
    end
  end

  private

  def cleanup_invalid_data
    @bathed_params = JobListing.cleanup_invalid_data(job_listing_params)
  end

  def check_for_subscription
    subscription = @company.subscription
    unless subscription and subscription.active?
      click_here_subscription = "<a href='/companies/#{@company.id}/subscriptions/new'>here</a>"
      notice = "You need to setup a subscription before you can add job listings, click #{click_here_subscription} to add a subscription"
      redirect_to company_job_listings_path(company_id: @company.id), notice: notice and return
    end
  end

  def get_job_listing
    @job_listing = JobListing.find(params[:id])
  end

  def job_listing_params
    params.require(:job_listing).permit(:job_title, :job_description, :estimated_work_hours, :salary_range_low,
    :salary_range_high, :vacation_days, :healthcare, :equity, :bonuses, :retirement, :fulltime, :remote, :hiring_time,
    :tech_stack_id, :location, :company_id, :active, :sponsorship_available, :dental, :vision, :life_insurance).tap do |whitelisted|
      JobListing::HASHABLE_PARAMS.each do |hash_param|
        whitelisted[hash_param] = params[:job_listing][hash_param] #our cleanup_invalid_data method handles invalid data here
      end
    end
  end

  def find_company
    @company = Company.find(params[:company_id])
  end

  def listing_response
    @job_listing.get_preference_defaults
    listing_response = { job_listing: @job_listing , tech_stacks: @company.collected_tech_stacks }
    respond_with listing_response
  end
end

File: app/assets/javascripts/app/controllers/job_listings_controller.js
jobListing.controller("JobListingCtrl", ['$scope', '$http', '$state', '$stateParams', '$location', 'JobListing', function($scope, $http, $state, $stateParams, $location, JobListing) {
  $scope.jobListing     = {}
  $scope.errors         = []
  $scope.success        = ''
  $scope.error_updating = ''

  if ($state.current.name == "edit") {
    JobListing.editListing.retrieveListing({
      job_listing_id: $stateParams['job_listing_id'],
      company_id: $stateParams['company_id'],
      action: 'retrieve_listing'
    }, function(response) {
       response.acceptable_languages = chunk(response.acceptable_languages, 6)
       response.practices = chunk(response.practices, 6)
       response.special_characteristics = chunk(response.special_characteristics, 6)
       response.experience_level = chunk(response.experience_level, 5)
       response.position_type = chunk(response.position_type, 5)
       response.perks = chunk(response.perks, 6)
       $scope.data = response
    });
    return $scope.update = function() {
      JobListing.editListing.update({
        company_id: $stateParams['company_id'],
        job_listing_id: $stateParams['job_listing_id'],
        action: 'update_listing',
        job_listing: {
          job_title: $scope.data.job_title,
          job_description: $scope.data.job_description,
          location: $scope.data.location,
          position_type: unchunk($scope.data.position_type),
          experience_level: unchunk($scope.data.experience_level),
          salary_range_low: $scope.data.salary_range_low,
          salary_range_high: $scope.data.salary_range_high,
          vacation_days: $scope.data.vacation_days,
          healthcare: $scope.data.healthcare,
          dental: $scope.data.dental,
          vision: $scope.data.vision,
          life_insurance: $scope.data.life_insurance,
          retirement: $scope.data.retirement,
          equity: $scope.data.equity,
          bonuses: $scope.data.bonuses,
          perks: unchunk($scope.data.perks),
          fulltime: $scope.data.fulltime,
          estimated_work_hours: $scope.data.estimated_work_hours,
          remote: $scope.data.remote,
          hiring_time: $scope.data.hiring_time,
          special_characteristics: unchunk($scope.data.special_characteristics),
          acceptable_languages: unchunk($scope.data.acceptable_languages),
          practices: unchunk($scope.data.practices)
        }},
        function(response) {
          $location.path("/companies/" + $stateParams['company_id'] + "/job_listings/" + $stateParams['job_listing_id'])
          return $scope.success;
        }
    )};
  } else if($state.current.name == "new") {
    JobListing.newListing.new({
      company_id: $stateParams['company_id'],
      action: 'new'
    }, function(response) {
       response.acceptable_languages = chunk(response.acceptable_languages, 6)
       response.practices = chunk(response.practices, 6)
       response.special_characteristics = chunk(response.special_characteristics, 6)
       response.experience_level = chunk(response.experience_level, 5)
       response.position_type = chunk(response.position_type, 5)
       response.perks = chunk(response.perks, 6)
       $scope.data = response
    });
    return $scope.create = function() {
      JobListing.createListing.create({
        company_id: $stateParams['company_id'],
        action: 'create',
        job_listing: {
          job_title: $scope.data.job_title,
          job_description: $scope.data.job_description,
          location: $scope.data.location,
          position_type: unchunk($scope.data.position_type),
          experience_level: unchunk($scope.data.experience_level),
          salary_range_low: $scope.data.salary_range_low,
          salary_range_high: $scope.data.salary_range_high,
          vacation_days: $scope.data.vacation_days,
          healthcare: $scope.data.healthcare,
          dental: $scope.data.dental,
          vision: $scope.data.vision,
          life_insurance: $scope.data.life_insurance,
          retirement: $scope.data.retirement,
          equity: $scope.data.equity,
          bonuses: $scope.data.bonuses,
          perks: unchunk($scope.data.perks),
          fulltime: $scope.data.fulltime,
          estimated_work_hours: $scope.data.estimated_work_hours,
          remote: $scope.data.remote,
          hiring_time: $scope.data.hiring_time,
          special_characteristics: unchunk($scope.data.special_characteristics),
          acceptable_languages: unchunk($scope.data.acceptable_languages),
          practices: unchunk($scope.data.practices)
        }},
        function(response) {
          $location.path("/companies/" + $stateParams['company_id'] + "/job_listings/")
          return $scope.success;
        }
    )};
  }
}]);
